Security/TwoFactor/Provider/TwoFactorProviderPreparationListener.php

```diff
@@ -6,25 +6,53 @@
 
 use Scheb\TwoFactorBundle\Security\Authentication\Token\TwoFactorToken;
 use Scheb\TwoFactorBundle\Security\TwoFactor\Event\TwoFactorAuthenticationEvent;
+use Symfony\Component\HttpFoundation\Session\SessionInterface;
+use Symfony\Component\HttpKernel\Event\FilterResponseEvent;
 
 class TwoFactorProviderPreparationListener
 {
+    private const CALLED_PROVIDERS_SESSION_KEY = '2fa_called_providers';
+
     /**
      * @var TwoFactorProviderRegistry
      */
     private $providerRegistry;
 
-    public function __construct(TwoFactorProviderRegistry $providerRegistry)
+    /**
+     * @var SessionInterface
+     */
+    private $session;
+
+    /**
+     * @var TwoFactorToken|null
+     */
+    private $twoFactorToken;
+
+    public function __construct(TwoFactorProviderRegistry $providerRegistry, SessionInterface $session)
     {
         $this->providerRegistry = $providerRegistry;
+        $this->session = $session;
     }
 
     public function onTwoFactorAuthenticationRequest(TwoFactorAuthenticationEvent $event)
     {
-        /** @var TwoFactorToken $token */
-        $token = $event->getToken();
-        $user = $token->getUser();
-        $providerName = $token->getCurrentTwoFactorProvider();
+        $this->twoFactorToken = $event->getToken();
+    }
+
+    public function onKernelResponse(FilterResponseEvent $event)
+    {
+        if (!$this->twoFactorToken instanceof TwoFactorToken) {
+            return;
+        }
+
+        $user = $this->twoFactorToken->getUser();
+        $providerName = $this->twoFactorToken->getCurrentTwoFactorProvider();
+        $calledProviders = $this->session->get(self::CALLED_PROVIDERS_SESSION_KEY, []);
```

Comment:
After thinking a bit more about this solution, I actually came across a scenario, when the solution wouldn't work as expected and could actually lead to a security issue.

Imagine you have two firewalls in your application, protecting separate parts of the application (e.g. public logic vs. admin area login). Each one has two-factor authentication enabled. You login in one of the firewalls, trigger the preparation for let's say the "email" provider. Now you also login in the other firewall. Information about prepared providers is shared in the session, so the second firewall will think the "email" provider is already prepared and not execute the preparation step.

So we need to scope this by firewall to make it secure.